Integration of CloudFoundry location and Vanilla entity

kiuby88 opened the pull request:
Prototype of integration of `CloudFoundryPaasLocation` and `VanillaCloudFoundryApplication` entity. It is just a prototype, for example, driver pattern is not included yet, the envs are not supported yet and the artifact downloading is hardcoded instead of using `DownloaderResolver`. Some sensores and new effectors are necessary, to modify the hardware profile (such as memory or instances).

Next steps:
- Adding env support (very easy)
- Adding driver pattern

kiuby88 committed:
Adding CloudFoundryPaasLocation management to VanillaCloudFoundryApplication entity

kiuby88 committed:
Adding unit test for VanillaCloudFoundryApplication entity

kiuby88 committed:
Adding live test for VanillCloudFoundryApplication entity

kiuby88 committed:
Adding yawl live test to deploy an VanillaCloudFoundryApplication entity using a a yaml blueprint

kiuby88 commented:
@andreaturli , would you take a look? 
I will be pleased to know your suggestions

andreaturli commented on src/test/resources/vanilla-cf-stadalone.yml:
```diff
@@ -17,23 +17,15 @@
-      disk_quota: 10
-      instances: 1
-
+location: pivotal-ws
```
I think it is better to leave the location as it was before, it is a useful how to define the location example

kiuby88 replied:
I can add a new file with the long  location description because this YAML is used by a a LiveTests which requires real pivotal-ws configuration

andreaturli commented on src/test/resources/vanilla-cf-stadalone.yml:
```diff
@@ -17,23 +17,15 @@
+    path: classpath://brooklyn-example-hello-world-sql-webapp-in-paas.war
+    buildpack: https://github.com/cloudfoundry/java-buildpack.git
+    domain: cfapps.io
+    memory: 512
```
is `memory` a location or an entity config?

andreaturli commented on src/test/resources/vanilla-cf-stadalone.yml:
```diff
@@ -17,23 +17,15 @@
+    buildpack: https://github.com/cloudfoundry/java-buildpack.git
+    domain: cfapps.io
+    memory: 512
+    disk_quota: 2048
```
is `disk_quota` a location or an entity config?

kiuby88 replied:
Currently, memory, disk and instances are entity config

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java:
```diff
@@ -117,7 +152,28 @@ protected void preStart(Collection<? extends Location> locations) {
+        Maybe<MachineLocation> ml = Locations.findUniqueMachineLocation(locations);
+        if (ml.isPresent()) return ml.get();
+
+        if (locations.isEmpty())
```
could it ever be empty?

kiuby88 replied:
Yes, it could be empty. Indeed I have added a test case to check the exception throwing

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java:
```diff
@@ -131,38 +187,86 @@ protected void customStart() {
+    }
+
+    protected boolean isApplicationDomainAvailable() {
+        boolean result;
```
maybe ?

``` java
boolean result = false;
try {
   result = HttpTool.getHttpStatusCode(applicationUrl) == HttpURLConnection.HTTP_OK;
} catch (Exception e) {
   log.warn(...);
}
return result;
```

kiuby88 replied:
+1

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/entity/VanillaCloudfoundryApplicationImpl.java:
```diff
@@ -235,10 +354,60 @@ public void restart() {
+        if (log.isDebugEnabled()) {
+            log.debug("waiting to ensure {} doesn't abort prematurely", this);
+        }
+        Duration startTimeout = getConfig(START_TIMEOUT);
```
instead of using a CoundownTimer, does the `cf-client` offers a `getApplicationStatus` ?

Please consider something like

``` java
Repeater.create("Wait until the application is running").until(new Callable<Boolean>() {
              public Boolean call() {
                 return isRunning();
             }})
     .limitIterationsTo(30)
     ...
    .run()
```

kiuby88 replied:
Let me try

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/utils/LocalResourcesDownloader.java:
```diff
@@ -0,0 +1,100 @@
+
+import com.google.common.base.Throwables;
+
+public class LocalResourcesDownloader {
```
add unit tests for `LocalResoureDowloader`

kiuby88 replied:
+1

kiuby88 commented:
Thanks for your review, @andreaturli.
Let me improve the code according to your suggestions

kiuby88 committed:
Improvements

Optimizing the code.
Avoiding CountdownTimer in waitForEntityStart.
New yaml to specify how a location can be configured
Testing LocalResourcesDownloader

kiuby88 commented:
@andreaturli I have addressing all your comments

andreaturli commented:
lgtm
